src/builtins.cpp

```diff
@@ -434,26 +434,155 @@ bool lStartsWithLLVM(llvm::StringRef name) {
+    // {ISPCTarget::avx512icl_x16, ISPCTarget::avx512skx_x16},
+    // {ISPCTarget::avx512skx_x16, ISPCTarget::generic_i1x16},
+
+    // {ISPCTarget::avx512spr_x32, ISPCTarget::avx512icl_x32},
+    // {ISPCTarget::avx512icl_x32, ISPCTarget::avx512skx_x32},
+    // {ISPCTarget::avx512skx_x32, ISPCTarget::generic_i1x32},
+
+    // TODO: needs generic_i1x64 be repaired first
+    // {ISPCTarget::avx512spr_x64, ISPCTarget::avx512icl_x64},
+    // {ISPCTarget::avx512icl_x64, ISPCTarget::avx512skx_x64},
+    // {ISPCTarget::avx512skx_x64, ISPCTarget::generic_i1x64},
+
+    // {ISPCTarget::avx2vnni_i32x4, ISPCTarget::avx2_i32x4},
+    // {ISPCTarget::avx2_i32x4, ISPCTarget::avx1_i32x4},
+    // {ISPCTarget::avx1_i32x4, ISPCTarget::sse4_i32x4},
+    // {ISPCTarget::sse4_i32x4, ISPCTarget::sse2_i32x4},
+    // {ISPCTarget::sse2_i32x4, ISPCTarget::generic_i32x4},
+
+    // {ISPCTarget::avx2vnni_i32x8, ISPCTarget::avx2_i32x8},
+    // {ISPCTarget::avx2_i32x8, ISPCTarget::avx1_i32x8},
+    // {ISPCTarget::avx1_i32x8, ISPCTarget::sse4_i32x8},
+    // {ISPCTarget::sse4_i32x8, ISPCTarget::sse2_i32x8},
+    // {ISPCTarget::sse2_i32x8, ISPCTarget::generic_i32x8},
+
+    // {ISPCTarget::avx2vnni_i32x16, ISPCTarget::avx2_i32x16},
+    // {ISPCTarget::avx2_i32x16, ISPCTarget::avx1_i32x16},
+    // {ISPCTarget::avx1_i32x16, ISPCTarget::generic_i32x16},
+
+    // {ISPCTarget::avx2_i16x16, ISPCTarget::sse4_i16x16},
+    // {ISPCTarget::sse4_i16x16, ISPCTarget::generic_i16x16},
+
+    // {ISPCTarget::sse4_i8x16, ISPCTarget::generic_i8x16},
+
+    // {ISPCTarget::avx2_i8x32, ISPCTarget::generic_i8x32},
+};
+// clang-format on
+
+// Traverse the target hierarchy to find the parent target.
+ISPCTarget GetParentTarget(ISPCTarget target) {
+    auto it = targetParentMap.find(target);
+    if (it != targetParentMap.end()) {
+        return it->second;
+    }
+    return ISPCTarget::none;
+}
+
+// Check if the module has used but unresolved symbols from the given set of symbols.
+bool lHasUnresolvedSymbols(llvm::Module *module, const llvm::StringSet<> &symbols) {
+    for (llvm::Function &F : module->functions()) {
+        auto name = F.getName().str();
+        if (F.isDeclaration() && !F.use_empty() && symbols.count(name)) {
+            return true;
+        }
+    }
+    return false;
+}
+
+void lFillTargetBuiltins(llvm::Module *bcModule, llvm::StringSet<> &targetBuiltins) {
+    for (llvm::Function &F : bcModule->functions()) {
         auto name = F.getName();
         if (!lStartsWithLLVM(name)) {
             targetBuiltins.insert(name);
         }
     }
+}
 
-    targetBCModule->setDataLayout(g->target->getDataLayout()->getStringRepresentation());
-    targetBCModule->setTargetTriple(module->getTargetTriple());
+llvm::Module *lGetTargetBCModule(ISPCTarget target) {
+    const BitcodeLib *targetBCLib = g->target_registry->getISPCTargetLib(target, g->target_os, g->target->getArch());
+    Assert(targetBCLib);
```

Comment:
Assert is good but I think error message would be beneficial here too especially in ispc-slim mode.
``` 
   if (!targetBCLib) {
        Error(SourcePos(), "Failed to get target bitcode library");
        return nullptr;
    }
```